RFC-0037-Interoperability-Standard-of-3rd-Backend-Integration-Mechanism.md

```diff
@@ -0,0 +1,91 @@
+
+Take Ascend NPU as an example:
+
+The Ascend NPU (torch_npu) successfully integrated with PyTorch based on the third-party device integration mechanism (PrivateUse1) in October 2023. Starting from November 2023, a set of daily tasks was established in the downstream code repository to test the functional compatibility between the latest PyTorch (main branch) and Ascend NPU (master branch) on a daily basis. As of March 2024, it has been running stably for approximately four months. The analysis results are shown in the table below:
+
+| Type | Description | Occurr Counts | Example PR |
+| :---: | :--- | :---: | :---: |
+| Refactoring | Turn Allocator::allocate into non-const, derived class’ override function is not modified. | 1 | [#120969](https://github.com/pytorch/pytorch/pull/120969) |
+| Refactoring | ​​Use DeviceIndex instead of int in CUDA wrappers, derived class’ override function is not modified. | 1 | [#119142](https://github.com/pytorch/pytorch/pull/119142) |
+| Refactoring | Move new trace utils from source to header, which leads to some symbols can’t be found. | 1 | [#114367](https://github.com/pytorch/pytorch/pull/114367/files) |
+| Refactoring | Migrate to getCvar* functions for env variable checking, which leads to function name can’t be found. | 1 | [#113797](https://github.com/pytorch/pytorch/pull/113797) |
+| New Features | Add support for new data types, data type assert fails. | 2 | [#107586](https://github.com/pytorch/pytorch/pull/107586), [#116594](https://github.com/pytorch/pytorch/pull/116594) |
+| New Features | Add function to materialize COW storages, which add a pure virtual function Allocator::copy_data, derived class didn’t implement this pure virtual function. | 2 | [#117053](https://github.com/pytorch/pytorch/pull/117053), [#113396](https://github.com/pytorch/pytorch/pull/113396) |
+
+Based on the above table, it is evident that upstream modifications do have a negative impact on out-of-tree third-party devices. These modifications directly affect the functionality of third-party devices and greatly increase the cost of adaptation and maintenance.
+
+
+### **Standardized Integration of Third-party Device**
+PyTorch is a complex framework with numerous features, making it a significant challenge for third-party devices to adapt to. The quality of integration methods directly impacts the stability and sustainability of the subsequent functionality of third-party backends. Therefore, standardizing the integration for third-party devices not only reduces integration barriers and time costs but also enhances compatibility and stability. This, in turn, solidifies PyTorch's leading position in the field of artificial intelligence frameworks.
+
+
+## **Proposed Implementation**
+![Architecture Outline](./RFC-0030-assets/3rd_backend_architecture.png)
```

Comment:
I love this diagram!
I think we want to do a couple clarifications here:
- For the middle case of XPU, showcase what still needs to be in core (allocator, stream, event, etc)
- Similarly for the out of tree case, would be curious to showcase what is in core and is providing the extension point being used.
- I would separate this PrivateUse1 device in core (that will look similar to XPU) and make it point to both out of core projects and point to the demo project in core (that is independent from the core integration).

Reply:
Yes.
I re-prepared the new diagram to show the architectural outline
![3rd_backend_architecture](https://github.com/pytorch/rfcs/assets/85333600/25523eff-f509-4f1f-abcf-058c129e485f)

Reply:
1. This is my mistake, I didn't notice that XPU has upstreamed some modules into pytorch, please refer to the latest diagram below.
2. The out of tree case is similar to that of IPU, only some necessary interfaces, logical branch codes and so on are integrated into the core. However, PrivateUse1 also has its own special features:
- Completeness: Support all torch functions as much as possible
- Universality: Provide flexibility to third-party devices as much as possible to shield differences between devices
3. As you commented below, it's probably best to keep the demo project as a PyTorch-organized project rather than keeping it in the tree.